File: backend/secure_research/citations.py
import re

from .models import CitationSource, SourceChunk


CITATION_RE = re.compile(r"\[(S\d+)\]")
CLAIM_SPLIT_RE = re.compile(r"(?<=[.!?])\s+|\n+")
WORD_RE = re.compile(r"[A-Za-z][A-Za-z'-]+")


def extract_citations(answer: str) -> list[str]:
    return list(dict.fromkeys(CITATION_RE.findall(answer)))


def validate_citations(
    answer: str, sources: list[SourceChunk]
) -> tuple[bool, list[str]]:
    cited = extract_citations(answer)
    allowed = {source.source_id for source in sources}
    markers_valid = bool(cited) and set(cited).issubset(allowed)
    claims = [
        claim.strip()
        for claim in CLAIM_SPLIT_RE.split(answer)
        if len(WORD_RE.findall(claim)) >= 4
    ]
    claims_cited = bool(claims) and all(CITATION_RE.search(claim) for claim in claims)
    return markers_valid and claims_cited, cited


def citation_sources(chunks: list[SourceChunk]) -> list[CitationSource]:
    return [
        CitationSource(
            source_id=chunk.source_id,
            page_id=chunk.page_id,
            page_title=chunk.page_title,
            page_url=chunk.page_url,
            heading_path=chunk.heading_path,
            block_ids=chunk.block_ids,
            score=chunk.score,
            excerpt=_excerpt(chunk.text),
        )
        for chunk in chunks
    ]


def _excerpt(text: str, max_chars: int = 280) -> str:
    compact = " ".join(text.split())
    if len(compact) <= max_chars:
        return compact
    return f"{compact[: max_chars - 1].rstrip()}…"

File: backend/server/src/retrieval/citations.ts
import type { CitationSource } from "@secure-research/contracts";
import type { SourceChunk } from "../types.ts";

const CITATION_RE = /\[(S\d+)\]/g;
const CLAIM_SPLIT_RE = /(?<=[.!?])\s+|\n+/;
const WORD_RE = /[A-Za-z][A-Za-z'-]+/g;

export type CitationValidation = {
  valid: boolean;
  citedSourceIds: string[];
  invalidSourceIds: string[];
  claimCount: number;
  uncitedClaimCount: number;
  hasCitations: boolean;
};

export function extractCitations(answer: string): string[] {
  return [...new Set([...answer.matchAll(CITATION_RE)].map((match) => match[1]))];
}

export function validateCitations(
  answer: string,
  sources: SourceChunk[],
): CitationValidation {
  const citedSourceIds = extractCitations(answer);
  const allowed = new Set(sources.map((source) => source.sourceId));
  const invalidSourceIds = citedSourceIds.filter((sourceId) => !allowed.has(sourceId));
  const claims = answer
    .split(CLAIM_SPLIT_RE)
    .map((claim) => claim.trim())
    .filter((claim) => (claim.match(WORD_RE) ?? []).length >= 4);
  const uncitedClaimCount = claims.filter(
    (claim) => !extractCitations(claim).some((sourceId) => allowed.has(sourceId)),
  ).length;
  const hasCitations = citedSourceIds.length > 0;

  return {
    valid:
      hasCitations &&
      invalidSourceIds.length === 0 &&
      claims.length > 0 &&
      uncitedClaimCount === 0,
    citedSourceIds,
    invalidSourceIds,
    claimCount: claims.length,
    uncitedClaimCount,
    hasCitations,
  };
}

export function citationSources(chunks: SourceChunk[]): CitationSource[] {
  return chunks.map((chunk) => ({
    sourceId: chunk.sourceId,
    pageId: chunk.pageId,
    pageTitle: chunk.pageTitle,
    pageUrl: chunk.pageUrl,
    headingPath: chunk.headingPath,
    blockIds: chunk.blockIds,
    score: chunk.score,
    excerpt: excerpt(chunk.text),
  }));
}

function excerpt(text: string, maxChars = 280): string {
  const compact = text.replace(/\s+/g, " ").trim();
  return compact.length <= maxChars
    ? compact
    : `${compact.slice(0, maxChars - 1).trimEnd()}…`;
}
